### PR TITLE
Add passivity module, ispassive function, and passivity_test. Introduces optional dependency cvxopt. 

### PR DESCRIPTION
**Pull request by Mark-Yeatman**
I made a discussion topic a couple of weeks ago about making a passivity module, @murrayrm suggested making a draft PR. Here it is!

### CODE REVIEW

**Comment by Mark-Yeatman** [comment]
I need to cut the formation of the LMI via symbolic variables for performance. Did a timing test on a size 20 state, input, output system, 99% of the time is spent constructing the LMI and converting to cvxopt form.

**Comment by Mark-Yeatman** [comment]
> Would it be possible to implement the matrix handling and solving with NumPy and SciPy functions only?

I'm about 60% sure cvxopt is necessary to solve the linear matrix inequality. Sympy and lmi_sdp can definitely be dropped.

Cvxopt only requires BLAS/LAPACK, which it seems that slycot also requires. So this passivity package could be setup to be optional?

File: control/exception.py [hunk 1]
```diff
@@ -84,3 +84,15 @@ def pandas_check():
...
+cvxopt_installed = None
+def cvxopt_check():
+    global cvxopt_installed
+    if cvxopt_installed is None:
+        try:
+            import cvxopt
+            cvxopt_installed = True
+        except:
+            cvxopt_installed = False
+    return cvxopt_installed
```
Comment (bnavigator) [inline control/exception.py:98]:
```suggestion
    return cvxopt_installed

```

File: .github/workflows/python-package-conda.yml [hunk 2]
```diff
@@ -13,6 +13,7 @@ jobs:
         python-version: [3.7, 3.9]
         slycot: ["", "conda"]
         pandas: [""]
+        cvxopt: ["conda"]
```
Comment (bnavigator) [inline .github/workflows/python-package-conda.yml:16]: Test available and not available cvxopt

```suggestion
        cvxopt: ["", "conda"]
```

File: control/passivity.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
+'''
+Author: Mark Yeatman  
+Date: May 15, 2022
+'''
+
+from . import statesp as ss
```
Comment (bnavigator) [inline control/passivity.py:6]: Please import the official API names, not the internal module file name as a name which is also a API function.

```suggestion
from .statesp import StateSpace
```

(You actually don't use anything directly)

File: control/tests/passivity_test.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+'''
+Author: Mark Yeatman  
+Date: May 30, 2022
+'''
+
+import numpy
+from control import ss, passivity
+from sympy import Matrix
```
Comment (bnavigator) [inline control/tests/passivity_test.py:8]:
```suggestion
```

**Review by bnavigator — COMMENTED** [review]
Thanks! Some initial review comments.

File: control/passivity.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import numpy as np
+import cvxopt as cvx
```
Comment (bnavigator) [inline control/passivity.py:8]: Make this import fail gracefully

File: control/passivity.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    Indicates if a linear time invarient system is passive
+
+    Constructs a linear matrix inequality and a feasibility optimization
+    such that is a solution exists, the system is passive.
+
+    The source for the algorithm is: 
+    McCourt, Michael J., and Panos J. Antsaklis. "Demonstrating passivity and dissipativity using computational methods." ISIS 8 (2013).
+    '''
+    if cvx is None:
+        raise ModuleNotFoundError
```
Comment (murrayrm) [inline control/passivity.py:24]: You might want to add a message to this exception so that the user knows what the need to do (eg, "cvxopt required for passivity module").

**Review by murrayrm — COMMENTED** [review]
A few comments below.  Also, it would be nice to add a unit test to make sure that if cvxopt is not installed that the proper exception is raised.

File: control/passivity.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+def is_passive(sys):
+    '''
+    Indicates if a linear time invarient system is passive
+
+    Constructs a linear matrix inequality and a feasibility optimization
+    such that is a solution exists, the system is passive.
+
+    The source for the algorithm is: 
+    McCourt, Michael J., and Panos J. Antsaklis. "Demonstrating passivity and dissipativity using computational methods." ISIS 8 (2013).
```
[the remark this answers is not in the document]
Comment (Mark-Yeatman) [reply control/passivity.py:21]: I have wrapped the line, but when I run "autopep8 --in-place passivity.py" it doesn't get corrected. Do you guys have a more aggressive formatter?

File: control/passivity.py (hunk 6 above)
Comment (Mark-Yeatman) [reply control/passivity.py:24]: Change taken.

File: control/passivity.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        )
+        return V
+
+    P = np.zeros_like(A)
+    matrix_list = []
+    state_space_size = sys.nstates
+    for i in range(0, state_space_size):
+        for j in range(0, state_space_size):
+            if j <= i:
+                P = P*0.0
```
[the remark this answers is not in the document]
Comment (Mark-Yeatman) [reply control/passivity.py:44]: I'm really not sure. We can change it.

File: .github/workflows/python-package-conda.yml [hunk 9]
```diff
@@ -13,6 +13,7 @@ jobs:
         python-version: [3.7, 3.9]
         slycot: ["", "conda"]
         pandas: [""]
+        cvxopt: ["", "conda"]
```
[the remark this answers is not in the document]
Comment (Mark-Yeatman) [reply .github/workflows/python-package-conda.yml:16]: I think I did this correctly in my last push? Would you mind giving a code suggestion for it if I didn't?

**Comment by Mark-Yeatman** [comment]
@bnavigator I believe I've resolved all the outstanding review comments?

File: control/passivity.py (hunk 6 above)
Comment (bnavigator) [reply control/passivity.py:27]: The output needs to be tied together. Print going to `stdout` might end up somewhere else than the backtrace (e.g. the exception and output captures by pytest):

```suggestion
        raise ModuleNotFoundError("cvxopt required for passivity module")
```

```
[ben@skylab:~]% python3 -c 'raise  ModuleNotFoundError("cvxopt required for passivity module")'                                            [0]
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: cvxopt required for passivity module
[ben@skylab:~]% python3 -c 'print("cvxopt required for passivity module"); raise ModuleNotFoundError'                                      [1]
cvxopt required for passivity module
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError
```

**Review by bnavigator — COMMENTED** [review]
One more review comment

File: control/passivity.py (hunk 6 above)
Comment (bnavigator) [reply control/passivity.py]: GitHub wouldn't let me suggest the removal of the print line in the same comment thread. Please remove the duplicate.

Comment (Mark-Yeatman) [reply control/passivity.py]: Will do

File: control/passivity.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+try:
+    import cvxopt as cvx
+except ImportError as e:
+    cvx = None
+
+
+def is_passive(sys):
+    '''
+    Indicates if a linear time invarient system is passive
```
Comment (JBannwarth) [inline control/passivity.py:16]: `invarient` -> `invariant`

I'm just lurking around so I'm not too sure about PR etiquette. Hopefully this is okay feedback.

Comment (Mark-Yeatman) [reply control/passivity.py:16]: I appreciate it!

**Comment by ilayn** [comment]
Instead of writing things in the standard SDP notation, you might want to use [CVXPY](https://www.cvxpy.org/examples/basic/sdp.html) notation (which acts like YALMIP) so you don't have to use the tiresome basis matrices. 

Also note that strictly proper systems cannot be checked via positive real lemma via strict inequalities due zero D + D.T block forcing the optimization problem to infeasibilty. You either have to use extended version or nudge D block with `epsilon*np.eye(m)`.

**Comment by Mark-Yeatman** [comment]
I would definitely be up for using cvxpy, but it looks like it has more dependencies than just using cvxopt (https://www.cvxpy.org/install/#install-from-source vs https://cvxopt.org/install/#building-and-installing-from-source)? That seems like an issue for a project maintainer. 

You're right about the D = 0 case, I will nudge it as you suggested.

**Comment by bnavigator** [comment]
CVXPY: I don't think it is necessary to employ a full domain specific language with all its dependencies for one hardcoded optimization problem.

File: control/passivity.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+import numpy as np
+
+try:
+    import cvxopt as cvx
+except ImportError as e:
+    cvx = None
+
+lmi_epsilon = 1e-12
+
+def is_passive(sys):
```
Comment (bnavigator) [inline control/passivity.py:15]: We have `(sys.)isdtime()`, `(sys.)isctime()`, and `(sys.)issiso()`. If we want to maintain a sense of consistency, this should be changed to `ispassive(sys)` and wrapped by a `sys.ispassive()` method for `StateSpace` systems, possibly even a wrapper for other LTI systems which realize them into `StateSpace` before calling the function in the passivity module.

`ispassive` would also be consistent with the naming scheme from [NumPy](https://numpy.org/doc/stable/reference/routines.logic.html) and [stdlib Python](https://docs.python.org/3/library/math.html#math.isclose)

File: control/tests/passivity_test.py [hunk 12]
```diff
@@ -15,10 +15,22 @@ def test_is_passive():
...
+    # happy path is passive
     assert(passivity.is_passive(sys))
 
+    # happy path not passive
     D = -D
     sys = ss(A, B, C, D)
 
     assert(not passivity.is_passive(sys))
 
+    #edge cases of D=0 boundary condition
```
Comment (bnavigator) [inline control/tests/passivity_test.py:27]: Please put into separate tests.

File: control/passivity.py [hunk 13]
```diff
@@ -10,6 +10,7 @@
 except ImportError as e:
     cvx = None
 
+lmi_epsilon = 1e-12
```
Comment (bnavigator) [inline control/passivity.py:13]: unused

File: control/passivity.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+try:
+    import cvxopt as cvx
+except ImportError as e:
+    cvx = None
+
+lmi_epsilon = 1e-12
+
+def is_passive(sys):
+    '''
+    Indicates if a linear time invariant system is passive
```
Comment (bnavigator) [inline control/passivity.py:17]: Please use [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) to indicate the input parameter and return type.

File: control/passivity.py (hunk 11 above)
Comment (Mark-Yeatman) [reply control/passivity.py:15]: Renaming the function for consistency sounds good.

I am planning on trying to add some sort of passivity related tools for systems beyond ones that are LTI (an LMI to check passivity was the lowest hanging fruit). In [Demonstrating Passivity and Dissipativity using Computational Methods](https://www3.nd.edu/~pantsakl/Publications/497-isis08%202013.pdf), there's methods for using sum-of-squares to compute passivity for polynomial nonlinear systems and I believe there's literature elsewhere for systems described by combinations of polynomials and trig functions. 

I think it is a good suggestion, but I'm not immediately sure what parent class to try to add a wrapper for ispassive() to.

File: control/tests/passivity_test.py [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+    # strictly proper
+    sys = ss(A, B, C, D)
+    assert(passivity.ispassive(sys))
+
+    # ill conditioned
+    A = A*1e12
+    sys = ss(A, B, C, D)
+    assert(passivity.ispassive(sys))
+
+    # different combinations of zero A,B,C,D are 0
```
[the remark this answers is not in the document]
Comment (Mark-Yeatman) [reply control/tests/passivity_test.py]: Got it! Thanks for working with me through this, sorry for some of the lazy commits.

**Comment by bnavigator** [comment]
I took the liberty of fixing the missing skips. I hope you don't mind.

File: control/lti.py [hunk 16]
```diff
@@ -201,6 +202,11 @@ def _dcgain(self, warn_infinite):
         else:
             return zeroresp
 
+    def ispassive(self):
+        # importing here prevents circular dependancy
+        from control.passivity import ispassive
+        ispassive(self)
```
Comment (bnavigator) [inline control/lti.py:208]:
```suggestion
        return(ispassive(self))
```

Comment (Mark-Yeatman) [reply control/lti.py]: Good catch!
